File: src/p0006/cpp/solution.cpp
class Solution {
public:
    string convert(const string &s, int numRows) {
        if (numRows <= 1) { return s; }
        string result;
        size_t rows = numRows;
        for (size_t k = 0; k < rows; k++) {
            for (size_t x = k; x < s.length();) {
                result.push_back(s[x]);
                x += 2 * (rows - 1);
                if (0 < k && k + 1 < rows) {
                    size_t y = x - k - k;
                    if (y < s.length()) {
                        result.push_back(s[y]);
                    }
                }
            }
        }
        return result;
    }
};

File: src/p0006/rust/solution.rs
impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        if num_rows <= 1 { return s; }
        let result = &mut Vec::new();
        let s = s.as_bytes();
        let rows = num_rows as usize;
        for k in 0 .. rows {
            let mut x = k;
            while x < s.len() {
                result.push(s[x]);
                x += 2 * (rows - 1);
                if 0 < k && k + 1 < rows {
                    let y = x - k - k;
                    if y < s.len() {
                        result.push(s[y]);
                    }
                }
            }
        }
        String::from_utf8_lossy(result).parse().unwrap()
    }
}
